rcl/test/rcl/test_wait.cpp

```diff
@@ -435,3 +435,256 @@ TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), guard_condition) {
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  rcl_guard_condition_t guard_cond_0 = rcl_get_zero_initialized_guard_condition();
+  ret = rcl_guard_condition_init(&guard_cond_0, rcl_guard_condition_get_default_options());
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  ret = rcl_wait_set_add_guard_condition(&wait_sets[0], &guard_cond_0);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  ret = rcl_trigger_guard_condition(&guard_cond_0);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  rcl_guard_condition_t guard_cond_1 = rcl_get_zero_initialized_guard_condition();
+  ret = rcl_guard_condition_init(&guard_cond_1, rcl_guard_condition_get_default_options());
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  ret = rcl_wait_set_add_guard_condition(&wait_sets[1], &guard_cond_1);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  ret = rcl_trigger_guard_condition(&guard_cond_1);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+    rcl_ret_t ret = rcl_guard_condition_fini(&guard_cond_0);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_guard_condition_fini(&guard_cond_1);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_wait_set_fini(&wait_sets[0]);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    ret = rcl_wait_set_fini(&wait_sets[1]);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  // Time spent during wait should be negligible.
+  int64_t timeout = 0;
+  std::chrono::steady_clock::time_point before_sc = std::chrono::steady_clock::now();
+  ret = rcl_wait_multiple(wait_sets, 2, timeout);
+  std::chrono::steady_clock::time_point after_sc = std::chrono::steady_clock::now();
+  // We don't expect a timeout here (since the guard condition had already been triggered)
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  int64_t diff = std::chrono::duration_cast<std::chrono::nanoseconds>(after_sc - before_sc).count();
+  EXPECT_LE(diff, TOLERANCE);
+}
+
+// Check that multiple wait sets are updated properly with timers
+TEST_F(CLASSNAME(WaitSetTestFixture, RMW_IMPLEMENTATION), wait_multiple_timers) {
+  rcl_clock_t clock;
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_ret_t ret = rcl_clock_init(RCL_STEADY_TIME, &clock, &allocator);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  const size_t kNumWaitSets = 3u;
+  rcl_wait_set_t wait_sets[kNumWaitSets];
+  rcl_guard_condition_t guard_conds[kNumWaitSets];
+  rcl_timer_t timers[kNumWaitSets];
+  for (size_t i = 0u; i < kNumWaitSets; ++i) {
+    wait_sets[i] = rcl_get_zero_initialized_wait_set();
+    ret = rcl_wait_set_init(&wait_sets[i], 0, 1, 1, 0, 0, rcl_get_default_allocator());
+    ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    // Add a dummy guard condition to avoid an error
```

Comment:
What error is the dummy guard condition avoiding?

Reply:
That is a good question. I copied this over from the existing test involving a timer, without it the test segfaults :fearful:

I dug into it a bit more, and it seems like the rmw guard condition count isn't being incremented in [rcl_wait_set_add_timer()](https://github.com/ros2/rcl/blob/f256da0a86e20c72c6babd394bb34dea697ecbbe/rcl/src/rcl/wait.c#L427). Adding a dummy guard condition appears to be a hack to get the count to increment...

I think it can be fixed easily by ensuring the rmw guard condition count is incremented when adding timers, but without understanding the reason for the current implementation I am hesitant to change it. @wjwwood, is this a bug or intentional?

Reply:
Nevermind, I see that the rmw guard condition count is increment for timers in `rcl_wait` as the comment says (my bad):

https://github.com/ros2/rcl/blob/f256da0a86e20c72c6babd394bb34dea697ecbbe/rcl/src/rcl/wait.c#L502

Setting the number of guard conditions to zero resolves the issue. I've also updated  the tests for `rcl_wait()` that originally confused me. See e40a26b